joblib/numpy_pickle.py

```diff
@@ -440,23 +487,35 @@ def load(filename, mmap_mode=None):
     object might not match the original pickled object. Note that if the
     file was saved with compression, the arrays cannot be memmaped.
     """
-    with open(filename, 'rb') as file_handle:
-        # We are careful to open the file handle early and keep it open to
-        # avoid race-conditions on renames. That said, if data are stored in
-        # companion files, moving the directory will create a race when
-        # joblib tries to access the companion files.
-        if _read_magic(file_handle) == _ZFILE_PREFIX:
-            if mmap_mode is not None:
-                warnings.warn('file "%(filename)s" appears to be a zip, '
-                              'ignoring mmap_mode "%(mmap_mode)s" flag passed'
-                              % locals(), Warning, stacklevel=2)
-            unpickler = ZipNumpyUnpickler(filename, file_handle=file_handle)
-        else:
-            unpickler = NumpyUnpickler(filename, file_handle=file_handle,
-                                       mmap_mode=mmap_mode)
+    # Backward compatibility with old compression strategy
+    if _first_bytes(filename).startswith(_ZFILE_PREFIX):
+        warnings.warn("The file '%s' has been generated with a joblib version "
+                      "less than 0.10. "
+                      "Please regenerate this pickle file." % filename,
+                      DeprecationWarning, stacklevel=2)
+        return load_compatibility(filename)
+
+    with closing(_read_fileobject(filename)) as f:
+        if (isinstance(f, (BinaryZlibFile, io.BufferedIOBase)) and
```

Comment:
Looks like BinaryZlibFile is a subclass of io.BufferedIOBase so you can just use `isintance(f, io.BufferedIOBase)`